[MRG+2] BUG Use epsilon threshold in `_samme_proba`

stephen-hoover opened the pull request:
The `ensemble.weight_boosting._samme_proba` function should preserve ordering of class probabilities within an example, but thresholding logic (to avoid taking the log of values <= 0) was disrupting this. Instead of thresholding `proba < 0` probabilities to 1e-5, threshold `proba < epsilon` to epsilon. This avoids the issue of, e.g., probability values of 0 becoming larger than values of 1e-7.

Add a unit test for `_samme_proba` which checks that probability ordering is unchanged. This test fails on the un-modified version of `_samme_proba`. 

Resolves issue #4944 .

stephen-hoover committed:
BUG Use epsilon threshold in `_samme_proba` and `_boost_real`

Instead of thresholding <0 probabilities to 1e-5, threshold <epsilon to epsilon. This avoids the issue of, e.g., probability values of 0 becoming larger than values of 1e-7.

Add a unit test for `_samme_proba` which checks that probability ordering is unchanged.

Resolves issue #4944 .

amueller commented:
LGTM

amueller commented:
ping @larsmans @glouppe @ndawe

amueller commented:
Actually, the same expression is at  https://github.com/stephen-hoover/scikit-learn/blob/threshold-boosting-samme_proba/sklearn/ensemble/weight_boosting.py#L515

stephen-hoover commented:
If I make that same threshold change there, I get a test failure:
```
======================================================================
FAIL: sklearn.tests.test_common.test_non_meta_estimators('AdaBoostClassifier', <class 'sklearn.ensemble.weight_boosting.AdaBoostClassifier'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/shoover/src/scikit-learn-env/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/shoover/src/scikit-learn/sklearn/utils/estimator_checks.py", line 742, in check_classifiers_train
    assert_greater(accuracy_score(y, y_pred), 0.83)
nose.proxy.AssertionError: 0.79000000000000004 not greater than 0.83
    """Fail immediately, with the given message."""
>>  raise self.failureException('0.79000000000000004 not greater than 0.83')
```
In that test, there's a step where a number of examples have zero probability for one of the classes. The algorithm seems to be sensitive to whether that 0 is replaced by 1e-5 or 2.22e-16. It'll take me a bit longer to figure out why.

amueller commented:
I am not very familiar with the algorithm, maybe the people who worked on it can say more.

ndawe commented:
Sorry for showing up late here. Taking a look...

ndawe commented:
Strange. The accuracy on that problem is 0.94 with 1e-5 and anything smaller than that gives 0.79. The default tree being boosted here is rather limited (max_depth=1 right?). The SAMME paper makes no mention of a threshold here AFAIK.

ndawe commented:
@amueller any specific reason for 0.83 in this test?

ndawe commented:
Nothing obviously wrong here as far as I can tell. _samme_proba is giving similar results for 1e-5 and 1e-6 and the dataset only has 100 samples, two features, and three classes. I just think the test is unstable. If I increase the dataset size to 1000 then the test passes with the threshold here as low as 2.22e-16.

stephen-hoover committed:
TST Increase samples for classifier tests

In `utils.estimator_checks.check_classifiers_train`, increase the number of examples used in testing from the default 100 to 300. I had spurious failures for the AdaBoostClassifier when using the defaults. Increasing to 300 examples makes the test more stable, and doesn't appear to significantly increase the time needed for the test.

stephen-hoover commented:
Increasing to 1000 samples in `check_classifiers_train` doesn't seem to have an adverse effect on test time (I didn't notice any change in time to run `tests.test_common`), so I went ahead and made that change.

ndawe commented:
+1 from me.

300 samples also seems to give a reasonable accuracy.

stephen-hoover commented:
It's not clear to me why that one test is failing. The error message is "sklearn.tests.test_common.test_non_meta_estimators('OneClassSVM', <class 'sklearn.svm.classes.OneClassSVM'>) ... Command exited with code -1073741819". That shouldn't be related to any of the changes I made. Any ideas?

ndawe commented:
All tests are passing on travis-ci. @amueller do you know what's going wrong here?

ogrisel commented:
The OneClassSVM failure is random and related to another bug: #4942. You can ignore it for this PR.

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -710,7 +710,7 @@ def check_classifiers_one_label(name, Classifier):
 
 
 def check_classifiers_train(name, Classifier):
-    X_m, y_m = make_blobs(random_state=0)
```
why is this needed? I would rather not make test run slower unless we have a really good reason.

stephen-hoover replied:
The test `sklearn.tests.test_common.test_non_meta_estimators` failed for the `AdaBoostClassifier` when using the default 100 samples. @ndawe thought that there were too few samples for the classifier to work stably, so I increased to 1000 samples. When I timed the tests, I didn't see an appreciable increase in run time.

I went back and timed the `sklearn.tests.test_common.test_non_meta_estimators` tests at 100, 300, and 1000 samples. It seems that changing to 1000 samples increases the mean time of those tests by ~1-2 seconds. I reduced to 300 samples, which also allows the `AdaBoostClassifier` to find an adequate solution, but doesn't increase execution time of the tests.

ogrisel commented:
Apart from my comment +1 as well.

ogrisel commented:
Sorry I had not read the full discussion. Will merge as soon as CI is green. Thanks.

ogrisel commented:
Thanks @stephen-hoover!

amueller commented:
thanks and sorry for vanishing mid-review.
